### PR TITLE
Load defaults once

### PR DESCRIPTION
**Pull request by arodier**
A work in progress, most of the changes are architectural, but he provisioning time should significantly decrease.
Actually in test

### CODE REVIEW

**Comment by arodier** [comment]
Now, the time dropped drastically, five minutes on my workstation, 10 minutes on Vultr, I don't know where is the bottleneck there.

**Comment by arodier** [comment]
The job is here: https://jenkins.homebox.space/job/ci-buster-nano-any-deploy/

**Comment by arodier** [comment]
I will remove all the individual playbooks, they are now useless.
```
Saturday 23 May 2020  16:59:31 +0000 (0:00:00.945)       0:10:05.408 ********** 
=============================================================================== 
system-prepare -------------------------------------------------------- 124.86s
dovecot --------------------------------------------------------------- 108.60s
certificates-all ------------------------------------------------------- 91.32s
postfix ---------------------------------------------------------------- 63.45s
ldap ------------------------------------------------------------------- 45.08s
sogo ------------------------------------------------------------------- 36.96s
dns-server-bind -------------------------------------------------------- 16.44s
system-post-install ---------------------------------------------------- 16.00s
packages --------------------------------------------------------------- 15.51s
opendkim --------------------------------------------------------------- 13.73s
remote-access ---------------------------------------------------------- 13.64s
opendmarc -------------------------------------------------------------- 11.03s
cert-perms ------------------------------------------------------------- 10.54s
load-defaults ----------------------------------------------------------- 5.03s
mta-sts ----------------------------------------------------------------- 4.56s
autoconfig -------------------------------------------------------------- 4.47s
nginx ------------------------------------------------------------------- 4.36s
autodiscover ------------------------------------------------------------ 3.31s
user-setup -------------------------------------------------------------- 2.34s
dns-server-bind-refresh ------------------------------------------------- 1.58s
well-known-services ----------------------------------------------------- 1.26s
ejabberd ---------------------------------------------------------------- 1.09s
borg-backup ------------------------------------------------------------- 0.98s
transmission ------------------------------------------------------------ 0.96s
roundcube --------------------------------------------------------------- 0.96s
setup ------------------------------------------------------------------- 0.85s
zabbix-server ----------------------------------------------------------- 0.84s
website-simple ---------------------------------------------------------- 0.67s
rspamd ------------------------------------------------------------------ 0.51s
privoxy ----------------------------------------------------------------- 0.47s
clamav ------------------------------------------------------------------ 0.46s
access-check ------------------------------------------------------------ 0.45s
roundcube-milters ------------------------------------------------------- 0.40s
backup-server ----------------------------------------------------------- 0.40s
sogo-milters ------------------------------------------------------------ 0.40s
external-ip-type -------------------------------------------------------- 0.34s
rspamd-web -------------------------------------------------------------- 0.26s
tor --------------------------------------------------------------------- 0.26s
access-report ----------------------------------------------------------- 0.25s
import-accounts --------------------------------------------------------- 0.24s
luks-remote ------------------------------------------------------------- 0.20s
dns-server-check-propagation -------------------------------------------- 0.09s
extra-certs ------------------------------------------------------------- 0.04s
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
total ----------------------------------------------------------------- 605.20s
```

File: install/playbooks/main-inline.yml [hunk 1]
```diff
@@ -7,72 +7,113 @@
...
     - role: nginx
+      tags: nginx
     - role: dns-server-bind
       when: bind.install
+      tags: bind
     - role: dns-server-check-propagation
       when: bind.install and bind.propagation.check
+      tags: bind
     - role: certificates-all
+      tags: ssl
```
Comment (fredericmoulins) [inline install/playbooks/main-inline.yml]: `certificates` for the tag?
Or `tls` if you want, but I see it more as about the certificates than the TLS configuration.
In any case, "SSL" is long gone now. :sweat_smile:

As it is, the `certificates-all` role should either be:
- tagged `always` and the different role tags be handled in the role (applied to the import_tasks create_one) if possible ;
- tagged with every tag needing a certificate, plus the `certificates` one.

Another way would be to keep several invocations with parameters, as it is today, each tagged with `certificates` and the tag of the role needing the certificate. It could be either in the play, or kept in role dependencies (sort of "hidden").

Just from looking at it, I think:
- the solution with the `always` tag should be avoided (for not handling the tags inside the role) ;
- I might prefer the certificate declaration to be visible (even if it might be a good fit for dependencies) ;
- the all-tags-on-certificates-all seems fine ;
- I might prefer the split version so that it is clear where and how to add a certificate for a new role added to the play (but then it might be a bit less readable with those vars declared in the play).

Well, one of the last two would be great, I think.

File: install/playbooks/main-inline.yml [hunk 2]
```diff
@@ -7,72 +7,113 @@
...
     - role: import-accounts
       when: mail.import.active and not system.devel
+      tags: import
     - role: extra-certs
       when: extra_certs is defined
+      tags: ssl
     - role: well-known-services
+      tags: nginx
     - role: system-post-install
+      tags: always
```
Comment (fredericmoulins) [inline install/playbooks/main-inline.yml]: I don't think the `system-post-install` needs to be run `always`.
It will disable the root access via SSH for example: not really expected if running a single tagged aspect.

File: install/roles/dovecot/files/access-check-blacklist.sh [hunk 3]
```diff
@@ -65,7 +65,7 @@ fi
 # Check if the IP address is blacklisted
 blacklisted="0"
 if [ -r "$blacklistFile" ]; then
-    blacklisted=$(grepcidr -x -c "$IP" "$blacklistFile")
+    blacklisted=$(echo "$IP" | grepcidr -x -c -f "$blacklistFile")
```
Comment (fredericmoulins) [inline install/roles/dovecot/files/access-check-blacklist.sh:68]: Maybe keep this change for another commit.

File: install/roles/dovecot/files/access-check-country.sh [hunk 4]
```diff
@@ -51,6 +51,7 @@ if [ "$CUSTOM_COUNTRIES_TRUST" = "1" ]; then
 fi
 
 if [ "$CUSTOM_COUNTRIES_TRUST_HOME" = "1" ]; then
+
```
Comment (fredericmoulins) [inline install/roles/dovecot/files/access-check-country.sh:54]: Maybe keep this change for another commit.

File: install/roles/dovecot/files/access-check-whitelist.sh [hunk 5]
```diff
@@ -53,7 +53,7 @@ fi
 # Check if the IP address is whitelisted
 whitelisted="0"
 if [ -r "$whitelistFile" ]; then
-    whitelisted=$(grepcidr -x -c "$IP" "$whitelistFile")
+    whitelisted=$(echo "$IP" | grepcidr -x -c -f "$whitelistFile")
```
Comment (fredericmoulins) [inline install/roles/dovecot/files/access-check-whitelist.sh:56]: Maybe keep this change for another commit.

File: install/roles/dovecot/files/dovecot-postlogin [hunk 6]
```diff
@@ -69,7 +69,7 @@ ALERT_ADDRESS=$(grep ALERT_ADDRESS /etc/homebox/access-check.conf | cut -f 2 -d
 # Let the post-login script to run external queries (for instance to IP address lookups)
 ALLOW_EXTERNAL_QUERIES=$(grep ALLOW_EXTERNAL_QUERIES /etc/homebox/access-check.conf | cut -f 2 -d =)
 
-# When importing emails, the remote command is always '127.0.0.1'
+# When importing emails, the remote IP is always '127.0.0.1'
```
Comment (fredericmoulins) [inline install/roles/dovecot/files/dovecot-postlogin:72]: Maybe keep this change for another commit.

File: install/roles/extra-certs/tasks/main.yml [hunk 7]
```diff
@@ -6,6 +6,6 @@
     apply:
       tags: certificates
   with_items:
-    - '{{ extra_certs }}'
+    - '{{ extra_certs | default([]) }}'
```
Comment (fredericmoulins) [inline install/roles/extra-certs/tasks/main.yml:9]: Maybe keep this change for another commit.

File: install/playbooks/main-inline.yml (hunk 1 above)
Comment (fredericmoulins) [reply install/playbooks/main-inline.yml]: Ok, outdated because the move to main.yml wasn't picked up. Applies to main.yml, now.

The tag on extra-certs should be `certficates` too. It is used already in the `extra-certs.yml` playbook.

File: install/playbooks/roles/ldap/tasks/main.yml [hunk 8]
```diff
@@ -100,6 +100,7 @@
   vars:
     cert_dir: 'ldap.{{ network.domain }}'
     entity_group: openldap
+    access_private_key: true
```
[the remark this answers is not in the document]
Comment (arodier) [reply install/playbooks/roles/ldap/tasks/main.yml:103]: I am planning to set the default to false. Otherwise, I noticed the CI failed before because the certificate file permissions were changed from the correct ones to the wrong ones by one of the next stage. I did not identified the guilty stage, but this was more a reinsurance, and a step in the direction to set the default to false.

File: install/playbooks/main-inline.yml (hunk 1 above)
Comment (arodier) [reply install/playbooks/main-inline.yml]: Yes, this is a huge work in progress. but I am happy with your suggestions, and even more that you are following ! I will continue in the afternoon

File: install/playbooks/main-inline.yml (hunk 2 above)
Comment (arodier) [reply install/playbooks/main-inline.yml]: In this case, I would prefer to fix the role, but having a role at the end will be useful. For instance to clean-up package cache, lock some partitions in r/o mode, etc...

File: install/roles/dovecot/files/access-check-blacklist.sh (hunk 3 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-blacklist.sh:68]: Yes, you are right. It is just painful to switch constantly to multiple branches.

File: install/roles/dovecot/files/access-check-country.sh (hunk 4 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-country.sh:54]: Yes, absolutely right.

File: install/roles/dovecot/files/access-check-whitelist.sh (hunk 5 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-whitelist.sh:56]: idem

File: install/roles/dovecot/files/dovecot-postlogin (hunk 6 above)
Comment (arodier) [reply install/roles/dovecot/files/dovecot-postlogin:72]: idem

File: install/roles/extra-certs/tasks/main.yml (hunk 7 above)
Comment (arodier) [reply install/roles/extra-certs/tasks/main.yml:9]: idem

Comment (arodier) [reply install/roles/extra-certs/tasks/main.yml:9]: I agree with the previous one, but this one seems needed to pass the build, or the load-defaults role does not work.

File: install/playbooks/main-inline.yml (hunk 1 above)
Comment (arodier) [reply install/playbooks/main-inline.yml]: in our case, ssl was more a tag chosen because it is short than anything else. As a short and quick tag, tls is fine too. Now, I had a look, and it seems we are using "cert" and "certificates", I will come back to this tomorrow or later.
The reason why I moved them all in one place is performance. This was the second role that was taking most of the time.

File: install/roles/dovecot/files/access-check-blacklist.sh (hunk 3 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-blacklist.sh]: Removed, it will be in another PR.

File: install/roles/dovecot/files/access-check-country.sh (hunk 4 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-country.sh]: Removed, it will be in another PR.

File: install/roles/dovecot/files/access-check-whitelist.sh (hunk 5 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-whitelist.sh]: Removed, it will be in another PR.

File: install/roles/dovecot/files/access-check-country.sh (hunk 4 above)
Comment (arodier) [reply install/roles/dovecot/files/access-check-country.sh]: It was just an accidental blank line...

File: install/roles/dovecot/files/dovecot-postlogin (hunk 6 above)
Comment (arodier) [reply install/roles/dovecot/files/dovecot-postlogin]: Removed, it will be in another PR.

**Comment by arodier** [comment]
So, the deployment is working, I am now running the tests, perhaps it will be another PR.

**Comment by arodier** [comment]
It is not yet ready for deployment, I found an annoying bug.
I am going to try to solve the whole testing.

**Comment by arodier** [comment]
The deployment is working, but not all the tests, which is normal, regarding the version change.
Please, have a look to it when you have the time, then we can decide to fix the errors one by one.
I am actually focusing on postfix.

File: install/playbooks/roles/ldap/tasks/main.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (fredericmoulins) [reply install/playbooks/roles/ldap/tasks/main.yml:103]: Ok, I understand. I agree with the goal of defaulting to false.
It's on my list to test which servers really require to have their user/group accessing their private key, because some (most) of them load it when they start and bind as root.

File: install/playbooks/main-inline.yml (hunk 1 above)
Comment (fredericmoulins) [reply install/playbooks/main-inline.yml]: `certificates` or `cert` is fine by me. I usually choose the longer forms, but that's just me.
The role is doing a lot of work, indeed.

**Comment by fredericmoulins** [comment]
> I am actually focusing on postfix.

Remember that, as it were, the dev branch was already not passing all the tests for postfix. The task "Check if the email is in the sent folder" in `check-simple-email.yml` did not pass. I couldn't figure out why the senders_bcc_map was not working. It should still be blocking. Apart from that one, everything else should run fine.
